Client/Frontend/Brave Notifications/BraveNotificationsPresenter.swift

```diff
@@ -0,0 +1,489 @@
+      if case .automatic(let interval) = notification.dismissPolicy {
+        setupTimeoutTimer(for: notification, interval: interval)
+      }
+      notificationView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+        animation.toValue = 0
+      }
+    default:
+      break
+    }
+  }
+  
+  // MARK: - Animations
+  
+  private func animateIn(adView: UIView) {
+    adView.layoutIfNeeded()
+    adView.layer.transform = CATransform3DMakeTranslation(0, -adView.bounds.size.height, 0)
+    
+    adView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+      animation.toValue = 0
+    }
+  }
+  
+  private func animateOut(adView: UIView, velocity: CGFloat? = nil, completion: @escaping () -> Void) {
+    adView.layoutIfNeeded()
+    let y = adView.frame.minY - view.safeAreaInsets.top - adView.transform.ty
+    
+    adView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+      animation.toValue = -(view.safeAreaInsets.top + y + adView.bounds.size.height)
+      if let velocity = velocity {
+        animation.velocity = velocity
+      }
+      animation.completionBlock = { _, _ in
+        completion()
+      }
+    }
+  }
+}
+
+extension BraveNotificationsPresenter {
+  class View: UIView {
+    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
+      // Only allow tapping the ad part of this VC
+      if let view = super.hitTest(point, with: event), view.superview is AdView {
+        return view
+      } else if let view = super.hitTest(point, with: event), view is WalletConnectionView {
+        return view
+      }
+      return nil
+    }
+  }
+}
+
+extension BraveNotificationsPresenter: UIGestureRecognizerDelegate {
+  
+  public func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
+    if let pan = gestureRecognizer as? UIPanGestureRecognizer {
+      let velocity = pan.velocity(in: pan.view)
+      if let adView = pan.view as? AdView, adView.swipeTranslation != 0 {
+        // dislike mode but swip vertically
+        if let notification = displayedNotifications.first(where: { $0.view == adView }) {
+          if case .automatic(let interval) = notification.dismissPolicy {
+            setupTimeoutTimer(for: notification, interval: interval)
+          }
+        }
+        // Vertical only and only if the user isn't in a swipe transform for a Rewards notification
+        return false
+      }
+      return abs(velocity.y) > abs(velocity.x)
+    }
+    return false
+  }
+}
+
+extension BraveNotificationsPresenter {
+  
+  /// Display a "My First Ad" on a presenting controller and be notified if they tap it
+  public static func displayFirstAd(on presentingController: UIViewController, completion: @escaping (BraveNotificationAction.Rewards, URL) -> Void) {
```

Comment:
Lets move this outside of the scope of the notifications presenter and onto the Ads notification handler or something

Reply:
move it to `RewardsNotification` class